client/state/sites/reducer.js

```diff
@@ -214,6 +215,32 @@ export function items( state = {}, action ) {
 
 			return state;
 		}
+
+		case InvitesActionTypes.RECEIVE_INVITE_ACCEPTED_SUCCESS: {
+			const { invite, data } = action;
+
+			if ( ! invite || ! data || ! data.sites ) {
+				return state;
+			}
+
+			if ( invite.role === 'follower' || invite.role === 'viewer' ) {
+				return state;
+			}
+
+			return reduce(
+				data.sites,
+				( memo, site ) => {
+					if ( memo === state ) {
+						memo = { ...state };
+					}
+
+					memo[ site.ID ] = pick( site, VALID_SITE_KEYS );
```

Comment:
The `SITES_RECEIVE` reducer is a bit smarter here: it checks with `isEqual` if the old and new site objects are deep-equal and if they are, doesn't update.

If you don't do this, all memoized selector values will be nuked and all components rerendered, because the two site objects are not `===` to each other.

Reply:
Ah, heh. That's why I didn't want to update all sites data, but just add the new site.

Reply:
Correction: I was too optimistic about the `SITES_RECEIVE` reducer. Because it adds all the received sites to a new empty `state` object, it doesn't keep the references to the old unchanged sites.

That's something that should be optimized, but that's a topic for a standalone PR.

By the way, receiving just single site is very rare in Calypso: there are just two usages of  the `SITE_RECEIVE` actions, and one of them can be removed because it does nothing :)